Bps 25

siandreev opened the pull request:
Coin and Explorer ether realisation added

siandreev committed:
Merge pull request #1 from dsx-tech/master

frok update

siandreev committed:
ethTransactions added

skoret commented on src/main/kotlin/dsx/bps/crypto/eth/EthCoin.kt:
```diff
@@ -0,0 +1,138 @@
+    override fun getTag(): Int? = Random.nextInt(0, Int.MAX_VALUE)
+
+    /**
+     * @param txid IxId object ( {hash : String, index : Int} )
```
IxId

siandreev committed:
ClientUnitTests fixed

skoret commented on src/main/kotlin/dsx/bps/crypto/eth/EthCoin.kt:
```diff
@@ -0,0 +1,138 @@
+
+            override fun fee(): BigDecimal {
+                return when {
+                    this.status() == TxStatus.VALIDATING -> hexToWei(ethTx.gasPrice)
```
предлагаю енумчики сравнивать конструкцией
```kotlin
when this.status() {
    TxStatus.VALIDATING -> ...
    ...
    else -> ...
}
```

skoret commented on src/main/kotlin/dsx/bps/crypto/eth/EthRpc.kt:
```diff
@@ -0,0 +1,78 @@
+    fun sendTransaction(
+        __from: String, __to: String, __gas: BigInteger = 90000.toBigInteger()
+        , __gasPrice: BigInteger = getGasPrice(), __value: BigDecimal, __data: String = "",
+        __nonce: BigInteger = getTransactionCount(__from)
```
нам правда необходимы двойные подчёркивания здесь?
по код-стайлу чуть-чуть: если аргументы в несколько строчек, то запятые в конце

skoret commented on src/main/kotlin/dsx/bps/rpc/JsonRpcHttpClient.kt:
```diff
@@ -40,11 +40,26 @@ abstract class JsonRpcHttpClient {
+                "id"     to id
+            ))
+        } */
+      //  else {
```
- [ ] ненужное — удаляем, для будущих доделок — `// TODO: <message>`

skoret commented on src/main/kotlin/dsx/bps/crypto/eth/datamodel/EthTxInfo.kt:
```diff
@@ -0,0 +1,20 @@
+    @SerializedName("data")
+    val data: String
+   // @SerializedName("nonce") //добавить потом когда нужна будет опция перезаписи транзакций
+    // val nonce: String //а лучше добавить с теми же остальными полями но в другом классе
```
`// TODO: <message in english>`

skoret commented on src/main/kotlin/dsx/bps/rpc/JsonRpcHttpClient.kt:
```diff
@@ -53,6 +68,7 @@ abstract class JsonRpcHttpClient {
         conn.doInput = true
         conn.doOutput = true
         conn.requestMethod = "POST"
+        conn.setRequestProperty("Content-Type", "application/json") //только для eth, говнокод
```
- [ ] вынести в EthRpc

для референса смотри в 
1) [BtcRpc#L14](https://github.com/dsx-tech/Blockchain-Payment-System/blob/master/src/main/kotlin/dsx/bps/crypto/btc/BtcRpc.kt#L14)
2) [JsonRpcHttpClient#L63](https://github.com/dsx-tech/Blockchain-Payment-System/blob/master/src/main/kotlin/dsx/bps/rpc/JsonRpcHttpClient.kt#L63)

skoret commented on src/main/kotlin/dsx/bps/core/datamodel/Currency.kt:
```diff
@@ -14,6 +15,10 @@ enum class Currency {
         override val coinConfigSpec = TrxConfig
+    },
+
+    ETH{
```
- [ ] пробелы между именем и скобкой

// для каждой валюты

phil-dolgolev commented on src/main/kotlin/dsx/bps/crypto/eth/EthCoin.kt:
```diff
@@ -0,0 +1,138 @@
+    }
+
+    /**
+     * @return account balance in ether (15 signs).
```
Может 15 конфирмейшенов?

phil-dolgolev commented on src/main/kotlin/dsx/bps/crypto/eth/EthCoin.kt:
```diff
@@ -0,0 +1,138 @@
+
+            override fun status(): TxStatus {
+                val conf = hexToBigInt(latestBlock.number) - hexToBigInt(ethTx.blockNumber)
+                return when {
```
Выглядит как чересчур фанатство when, тут даже старый добрый тернарный оператор отлично зайдет :)

phil-dolgolev commented on src/main/kotlin/dsx/bps/crypto/eth/EthCoin.kt:
```diff
@@ -0,0 +1,138 @@
+        val latestBlock = rpc.getLatestBlock()
+
+        return object : Tx {
+            override fun currency() = Currency.ETH
```
Я правильно понимаю, что здесь создается анонимный класс каждый раз? 
Довольно странная штука, почему не запилен именно класс для этого дела, реализующий интерфейс?

phil-dolgolev commented on src/main/kotlin/dsx/bps/crypto/eth/EthCoin.kt:
```diff
@@ -0,0 +1,138 @@
+     */
+    override fun sendPayment(amount: BigDecimal, address: String, tag: Int?): Tx {
+        rpc.unlockAccount(accountAddress, privateKey)
+        val txHash = rpc.sendTransaction(__from = accountAddress, __to = address, __value = amount)
```
На самом деле это очень упрощенная реализация, которую точно надо будет менять

1) Nonce нужно контролировать на своей стороне
2) Нужно рассматривать вариант, когда sendTransaction отвалился с ошибкой, ибо по факту транзакция может уйти в сеть, да только мы не узнаем об этом, что приведет к даблспенду. 
3) Идеально - сделать подпись транзакции в приложении, а не разлачивать в ноде кошелек. И вообще, надо идти к тому, чтоб в ноде никакого контексте не было и мы на него не пологались, кошельки храним отдельно

phil-dolgolev commented on src/main/kotlin/dsx/bps/crypto/eth/EthConverter.kt:
```diff
@@ -0,0 +1,21 @@
+import java.math.BigDecimal
+import java.math.BigInteger
+
+fun hexToEth(hex: String): BigDecimal {
```
Web3j  как раз позволит не реализовывать такие штуки самому с нуля

phil-dolgolev commented on src/main/kotlin/dsx/bps/crypto/eth/EthExplorer.kt:
```diff
@@ -0,0 +1,36 @@
+
+    override fun explore() {
+        var last = coin.getLatestBlock()
+        viewed.add(last.hash)
```
При перезапусках мы не заполняем историю? Это надо как-то обойти

phil-dolgolev commented on src/main/kotlin/dsx/bps/crypto/eth/EthExplorer.kt:
```diff
@@ -0,0 +1,36 @@
+                            val tx = coin.constructTx(it)
+                            emitter.onNext(tx)
+                        }
+                    viewed.add(new.hash)
```
Добавление сразу во viewed означает, что если мы свалимся с ошибкой дальше, то у нас потенциально возникает пропуск блоков, или я что-то не так понял?

phil-dolgolev commented on src/main/kotlin/dsx/bps/crypto/eth/EthRpc.kt:
```diff
@@ -0,0 +1,78 @@
+import java.math.BigInteger
+
+
+class EthRpc(url: String) : JsonRpcHttpClient(url) {
```
Нужно рассмотреть замену на Web3j

phil-dolgolev commented on src/main/kotlin/dsx/bps/rpc/JsonRpcHttpClient.kt:
```diff
@@ -40,11 +40,26 @@ abstract class JsonRpcHttpClient {
-        ))
+        val json : String
+      /*  if (params.contains("\"jsonrpc\":\"2.0\"")) {
+            var cleanedParams = params.filter { x->x != "\"jsonrpc\":\"2.0\""} //говнокод
```
Отличный PR :D
Удаляй "говнокод" :)

phil-dolgolev commented on src/test/kotlin/dsx/bps/crypto/eth/EthRpcTest.kt:
```diff
@@ -0,0 +1,86 @@
+import org.junit.jupiter.api.Test
+
+@Disabled
+internal class EthRpcTest {
```
Вот для этого нам и нужны testcontainers, чтобы эти тесты на только руками можно было запускать на определенном локальном энвайроменте, но и на CI

siandreev committed:
Сделан переход на web3j.
Изменена работа с аккаунтами, но необходимо доработать сохранение паролей для новых адресов (пока что у всех одинаковый, взятый из конфигов).
Разделены процессы подписания и отправки транзакций.
Nonce теперь контролируется вручную, но необходимо обеспечить его безопасное изменение.
Прочие замечания поправлены.

skoret commented on build.gradle:
```diff
@@ -9,20 +10,37 @@ repositories {
     implementation "com.uchuhimo:konf:0.20.0"
     implementation "com.uchuhimo:konf-yaml:0.20.0"
 
+
```
почисти, пожалуйста, новые лишние пустые линии

skoret commented on src/main/kotlin/dsx/bps/crypto/common/Coin.kt:
```diff
@@ -13,7 +12,7 @@ abstract class Coin {
     abstract val config: Config
 
-    protected abstract val rpc: JsonRpcHttpClient
+    protected abstract val rpc: Any
```
возможно, стоит завести интерфейс Connector (или что-то в этом роде)
пусть даже пустой без методов
здесь, в Coin, rpc будет типа этого интерфейса
а дальше и JsonRpcHttpClient, и EthRpc будут его "реализовывать"
так мы явно будем сигнализировать, что это коннектор к ноде, а не что-Any-угодно

skoret commented on src/main/kotlin/dsx/bps/crypto/eth/EthCoin.kt:
```diff
@@ -69,52 +81,24 @@ class EthCoin : Coin {
-            }
-        }
+    fun constructTx(ethTx: Transaction): Tx {
+        return EthTxStandart(ethTx, rpc,confirmations)
```
в ответ @phil-dolgolev на коммент, относительно создания объекта, а не отдельного класса

в таких местах создавались именно объекты, а не классы, т.к. для конвертации специфичных EthTx (и для других валют аналогично) в обобщённый вид Tx необходима дополнительная информация, которой может обладать CoinClient
нр, те же вызовы rpc клиента, чтобы понять, сколько у нас подтверждений
если писать реализацию интерфейса Tx  для каждого класса, то это должен быть data class
который из себя — исключительно данные, без логики
отдавать в такой класс ссылку на rpc, да ещё и с публичным доступом, совсем не хочется

skoret commented on src/main/kotlin/dsx/bps/crypto/eth/EthCoin.kt:
```diff
@@ -123,16 +107,19 @@ class EthCoin : Coin {
+        val credentials = WalletUtils.loadCredentials(password, pathToWallet)
+        val signedTransaction = rpc.signTransaction(rawTransaction, credentials)
+        val resultHash = rpc.sendTransaction(signedTransaction)
+        nonce ++ // разработать безопасную работу с nonce
```
ещё раз: такие штуки помечаем как `// TODO: <comment in english>`

skoret commented on src/main/kotlin/dsx/bps/crypto/eth/EthCoin.kt:
```diff
@@ -123,16 +107,19 @@ class EthCoin : Coin {
     }
 
-    fun getLatestBlock(): EthBlock {
+    fun getLatestBlock(): org.web3j.protocol.core.methods.response.EthBlock.Block {
```
одни типы (нр, Transaction из web3j) импортированы
другие используются так, как здесь
стремимся к однообразию → юзаем импорты как и везде
**upd:** в остальных файлах в пулл реквесте тоже

skoret commented on src/main/kotlin/dsx/bps/crypto/eth/EthRpc.kt:
```diff
@@ -1,78 +1,126 @@
+    }
+
+    fun createRawTransaction(nonce : BigInteger, gasPrice : BigInteger = getGasPrice(),
+                             gasLimit : BigInteger = 90000.toBigInteger(), toAddress : String,
```
магические числа лучше выносить в константы с поясняющим их названием

skoret reviewed (changes requested):
- Отвечаем и резолвим комменты на уже исправленные замечания, чтобы трекать прогресс
- Если исправления для какого-то замечания не нужны, объясняем/спрашиваем/обсуждаем так же в комментах
- Когда-нибудь я обязательно залью код-стайл конфиг в репозиторий, а пока перед коммитом/пушем/пулл-реквестом прошу пробегаться по всему новому и производить косметический рефакторинг. Ориентироваться можно на уже готовые модули (особое внимание на необходимые/лишние пробелы, переносы строк, также подсказки IDE часто очень полезны)

siandreev committed:
Исправлены замечания по код-стайлу.

phil-dolgolev commented:
Не актуален
